laspy/copc.py

```diff
@@ -0,0 +1,863 @@
+        nodes_to_read = sorted(nodes_to_read, key=attrgetter("offset"))
+        grouped_nodes: List[List[OctreeNode]] = []
+        current_group: List[OctreeNode] = []
+        last_node_end = nodes_to_read[0].offset
+        for node in nodes_to_read:
+            if node.offset == last_node_end:
+                current_group.append(node)
+                last_node_end += node.byte_size
+            else:
+                grouped_nodes.append(current_group)
+                current_group = [node]
+                last_node_end = node.offset + node.byte_size
+        if current_group:
+            grouped_nodes.append(current_group)
+
+        compressed_bytes, num_points, chunk_table = self._fetch_all_chunks(
+            grouped_nodes
+        )
+        points_array = np.zeros(
+            num_points * self.header.point_format.size, dtype=np.uint8
+        )
+
+        lazrs.decompress_points_with_chunk_table(
+            compressed_bytes, self.laszip_vlr.record_data, points_array, chunk_table
+        )
+        r = PackedPointRecord.from_buffer(points_array, self.header.point_format)
+        points = ScaleAwarePointRecord(
+            r.array, r.point_format, self.header.scales, self.header.offsets
+        )
+
+        return points
+
+    def _fetch_all_chunks(
+        self, grouped_nodes: List[List[OctreeNode]]
+    ) -> Tuple[bytearray, int, List[Tuple[int, int]]]:
+
+        num_points = 0
+        num_compressed_bytes = 0
+        chunk_table: List[Tuple[int, int]] = []
+        byte_queries: List[Tuple[int, int]] = []
+        for group in grouped_nodes:
+            num_compressed_group_bytes = 0
+            for node in group:
+                chunk_table.append((node.point_count, node.byte_size))
+                num_compressed_group_bytes += node.byte_size
+                num_points += node.point_count
+
+            num_compressed_bytes += num_compressed_group_bytes
+            byte_queries.append((group[0].offset, num_compressed_group_bytes))
+
+        compressed_bytes = bytearray(num_compressed_bytes)
+
+        if isinstance(self.source, HttpRangeStream):
+            http_queue_strategy(
+                self.source, byte_queries, compressed_bytes, self.http_num_threads
+            )
+            # http_thread_executor_strategy(self.source, byte_queries, compressed_bytes, self.http_num_threads)
+
+        elif hasattr(self.source, "readinto"):
+            citer = ChunkIter(np.frombuffer(compressed_bytes, np.uint8))
+            for offset, size in byte_queries:
+                self.source.seek(offset)
+                cc = citer.next(size)
+                self.source.readinto(cc)
+        else:
+            citer = ChunkIter(np.frombuffer(compressed_bytes, np.uint8))
+            for offset, size in byte_queries:
+                self.source.seek(offset)
+                cc = citer.next(size)
+                cc[:] = np.frombuffer(self.source.read(size), np.uint8)
```

Comment:
could be avoid all those `np.from_buffer` cast by using [memoryviews](https://docs.python.org/3/library/stdtypes.html#memoryview)